File: src/server/room.ts
import { ChangshaMahjong } from "../engine/mahjong/game.js";
import { tileName } from "../engine/mahjong/tiles.js";
import { ShaoyangPaohuzi } from "../engine/paohuzi/game.js";
import { phzName } from "../engine/paohuzi/tiles.js";
import type {
  GameType,
  OpOption,
  PublicRoomState,
  SeatPublic,
} from "../shared/protocol.js";

export interface SeatConn {
  seat: number;
  nick: string;
  ready: boolean;
  isBot: boolean;
  send: (data: unknown) => void;
}

function randomRoomId(): string {
  return Math.random().toString(36).slice(2, 8).toUpperCase();
}

export class Room {
  id: string;
  gameType: GameType;
  seats: (SeatConn | null)[];
  game: ChangshaMahjong | ShaoyangPaohuzi;
  started = false;

  constructor(gameType: GameType, id = randomRoomId()) {
    this.id = id;
    this.gameType = gameType;
    const n = gameType === "changsha_mj" ? 4 : 3;
    this.seats = Array.from({ length: n }, () => null);
    this.game =
      gameType === "changsha_mj" ? new ChangshaMahjong() : new ShaoyangPaohuzi();
  }

  get playerCount() {
    return this.seats.length;
  }

  addPlayer(nick: string, send: (data: unknown) => void, isBot = false): number {
    const seat = this.seats.findIndex((s) => s === null);
    if (seat < 0) throw new Error("房间已满");
    this.seats[seat] = {
      seat,
      nick: nick || (isBot ? `机器人${seat}` : `玩家${seat}`),
      ready: isBot,
      isBot,
      send,
    };
    return seat;
  }

  fillBots() {
    while (this.seats.some((s) => s === null)) {
      this.addPlayer("", () => {}, true);
    }
  }

  setReady(seat: number) {
    const s = this.seats[seat];
    if (s) s.ready = true;
  }

  tryStart() {
    if (this.started) return;
    if (this.seats.some((s) => !s || !s.ready)) return;
    this.started = true;
    this.game.start();
    this.broadcastState();
    this.kickBots();
  }

  /** 若轮到机器人，自动行动 */
  kickBots() {
    const maxSteps = 24;
    for (let i = 0; i < maxSteps; i++) {
      if (this.game.phase === "finished") return;
      let acted = false;
      for (const s of this.seats) {
        if (!s?.isBot) continue;
        const ops = this.game.getOpsFor(s.seat);
        if (!ops.length) continue;
        this.botAct(s.seat, ops);
        acted = true;
        break;
      }
      if (!acted) break;
    }
  }

  botAct(seat: number, ops: OpOption[]) {
    // 优先级：胡 > 杠/提/跑 > 碰 > 吃 > 出牌/过（响应阶段必须立刻表态，避免卡桌）
    const prefer = (names: string[]) => ops.find((o) => names.includes(o.action));
    const win = prefer(["zimo", "hu"]);
    if (win) {
      this.applyAction(seat, win.action, { tile: win.tile, tiles: win.tiles });
      return;
    }
    const gang = prefer(["ming_gang", "an_gang", "bu_gang", "ti", "pao"]);
    if (gang) {
      this.applyAction(seat, gang.action, { tile: gang.tile, tiles: gang.tiles });
      return;
    }
    const peng = prefer(["peng"]);
    if (peng && Math.random() > 0.45) {
      this.applyAction(seat, peng.action, { tile: peng.tile });
      return;
    }
    const chi = prefer(["chi"]);
    if (chi && Math.random() > 0.55) {
      this.applyAction(seat, chi.action, { tiles: chi.tiles });
      return;
    }
    if (ops.some((o) => o.action === "pass")) {
      this.applyAction(seat, "pass", {});
      return;
    }
    if (ops.some((o) => o.action === "discard")) {
      const hand =
        this.gameType === "changsha_mj"
          ? (this.game as ChangshaMahjong).players[seat].hand
          : (this.game as ShaoyangPaohuzi).players[seat].hand;
      const tile = hand[Math.floor(Math.random() * hand.length)];
      this.applyAction(seat, "discard", { tile });
    }
  }

  applyAction(
    seat: number,
    action: string,
    payload: { tile?: number; tiles?: number[] },
  ): string | null {
    const err = this.game.apply(seat, action, payload);
    if (err) return err;
    this.broadcastState();
    // 下一局：finished 后若全员准备可再开 —— Demo 自动准备 bots，人类点准备
    if (this.game.phase === "finished") {
      for (const s of this.seats) {
        if (s?.isBot) s.ready = true;
        else if (s) s.ready = false;
      }
      this.started = false;
    } else {
      queueMicrotask(() => this.kickBots());
    }
    return null;
  }

  nextRound() {
    if (this.game.phase !== "finished") return;
    if (this.seats.some((s) => !s?.ready)) return;
    this.started = true;
    this.game.start();
    this.broadcastState();
    this.kickBots();
  }

  broadcastState() {
    for (const s of this.seats) {
      if (!s || s.isBot) continue;
      s.send({ type: "state", state: this.buildPublic(s.seat) });
    }
  }

  buildPublic(viewerSeat: number): PublicRoomState {
    const g = this.game;
    const seats: SeatPublic[] = this.seats.map((s, i) => {
      const p =
        this.gameType === "changsha_mj"
          ? (g as ChangshaMahjong).players[i]
          : (g as ShaoyangPaohuzi).players[i];
      return {
        seat: i,
        nick: s?.nick ?? `空位${i}`,
        ready: s?.ready ?? false,
        isBot: s?.isBot ?? false,
        handCount: p.hand.length,
        melds: g.toPublicMelds(i),
        discards: p.discards,
        score: p.score,
        hand: i === viewerSeat ? [...p.hand] : undefined,
      };
    });

    const nameFn = this.gameType === "changsha_mj" ? tileName : phzName;
    void nameFn;

    return {
      roomId: this.id,
      gameType: this.gameType,
      phase: g.phase,
      seats,
      wallCount:
        this.gameType === "changsha_mj"
          ? (g as ChangshaMahjong).wall.length
          : (g as ShaoyangPaohuzi).wall.length,
      currentSeat: g.currentSeat,
      lastDiscard: g.lastDiscard ?? undefined,
      availableOps: g.getOpsFor(viewerSeat),
      message: g.message,
      round: g.round,
      settle: g.settle,
    };
  }
}

export class RoomManager {
  rooms = new Map<string, Room>();

  create(gameType: GameType, nick: string, send: (d: unknown) => void) {
    const room = new Room(gameType);
    this.rooms.set(room.id, room);
    const seat = room.addPlayer(nick, send, false);
    return { room, seat };
  }

  join(roomId: string, nick: string, send: (d: unknown) => void) {
    const room = this.rooms.get(roomId.toUpperCase());
    if (!room) throw new Error("房间不存在");
    if (room.started) throw new Error("对局已开始");
    const seat = room.addPlayer(nick, send, false);
    return { room, seat };
  }
}

File: server/lualib/room.lua
--[[ 房间逻辑：座位 / 机器人 / 广播快照 ]]
local Catalog = require "game_catalog"

local Room = {}
Room.__index = Room

local function rand_id()
  local chars = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
  local t = {}
  for _ = 1, 6 do
    local j = math.random(#chars)
    t[#t + 1] = chars:sub(j, j)
  end
  return table.concat(t)
end

function Room.new(game_type, id)
  local self = setmetatable({}, Room)
  self.id = id or rand_id()
  local meta = Catalog.get(game_type)
  self.game_type = meta.id
  self.started = false
  self.game = meta.factory()
  self.n = meta.seats
  self.seats = {}
  for i = 0, self.n - 1 do self.seats[i] = nil end
  return self
end

function Room:add_player(nick, agent, is_bot)
  local seat
  for i = 0, self.n - 1 do
    if self.seats[i] == nil then seat = i; break end
  end
  if seat == nil then return nil, "房间已满" end
  self.seats[seat] = {
    seat = seat,
    nick = nick or (is_bot and ("机器人" .. seat) or ("玩家" .. seat)),
    ready = is_bot and true or false,
    is_bot = is_bot and true or false,
    agent = agent,
  }
  return seat
end

function Room:fill_bots()
  for i = 0, self.n - 1 do
    if self.seats[i] == nil then
      self:add_player(nil, nil, true)
    end
  end
end

function Room:set_ready(seat)
  if self.seats[seat] then self.seats[seat].ready = true end
end

function Room:all_ready()
  for i = 0, self.n - 1 do
    local s = self.seats[i]
    if not s or not s.ready then return false end
  end
  return true
end

function Room:build_public(viewer)
  local g = self.game
  local seats = {}
  for i = 0, self.n - 1 do
    local s = self.seats[i]
    local p = g.players[i]
    local hand
    if i == viewer then
      hand = {}
      for _, t in ipairs(p.hand) do hand[#hand + 1] = t end
    end
    seats[#seats + 1] = {
      seat = i,
      nick = s and s.nick or ("空位" .. i),
      ready = s and s.ready or false,
      isBot = s and s.is_bot or false,
      handCount = #p.hand,
      melds = g:public_melds(i),
      discards = p.discards,
      score = p.score,
      hand = hand,
    }
  end
  local lastDiscard = g.last_discard
  if not lastDiscard and g.last_play and g.last_play.cards and #g.last_play.cards > 0 then
    lastDiscard = { seat = g.last_play.seat, tile = g.last_play.cards[#g.last_play.cards] }
  end
  return {
    roomId = self.id,
    gameType = self.game_type,
    phase = g.phase,
    seats = seats,
    wallCount = g.wall and #g.wall or (g.bottom and #g.bottom or 0),
    currentSeat = g.current_seat,
    lastDiscard = lastDiscard,
    lastPlayCards = g.last_play and g.last_play.cards or nil,
    availableOps = g:get_ops(viewer),
    message = g.message,
    round = g.round,
    settle = g.settle,
    landlord = g.landlord,
  }
end

function Room:broadcast(send_fn)
  for i = 0, self.n - 1 do
    local s = self.seats[i]
    if s and s.agent and not s.is_bot then
      send_fn(s.agent, { type = "state", state = self:build_public(i) })
    end
  end
end

function Room:bot_act(seat, ops)
  local function prefer(names)
    for _, op in ipairs(ops) do
      for _, n in ipairs(names) do
        if op.action == n then return op end
      end
    end
  end

  -- 斗地主叫分
  if self.game_type == "doudizhu" and self.game.phase == "bidding" then
    local bids = {}
    for _, op in ipairs(ops) do
      if op.action:sub(1, 4) == "bid_" then bids[#bids + 1] = op end
    end
    if #bids > 0 then
      if math.random() < 0.35 then
        return "bid_0", {}
      end
      local best = bids[#bids]
      return best.action, { tile = best.tile }
    end
  end

  -- 斗地主 / 跑得快出牌
  if (self.game_type == "doudizhu" or self.game_type == "paodekuai") and self.game.phase == "playing" then
    local Engine = Catalog.get(self.game_type).module
    local hand = self.game.players[seat].hand
    if prefer({ "pass" }) and self.game.last_play and self.game.last_play.seat ~= seat then
      local prev = self.game.last_play.pattern
      if prev and prev.kind == "single" then
        for _, c in ipairs(hand) do
          local pat = Engine.parse_pattern({ c })
          if Engine.beats(pat, prev) then
            return "play", { tiles = { c } }
          end
        end
      end
      return "pass", {}
    end
    if prefer({ "play" }) or prefer({ "discard" }) then
      -- 跑得快首出必须带红桃3
      if self.game_type == "paodekuai" and self.game.first_play then
        local h3 = 13
        for _, c in ipairs(hand) do
          if c == h3 then return "play", { tiles = { c } } end
        end
      end
      local c = hand[1]
      return "play", { tiles = { c } }
    end
  end

  local win = prefer({ "zimo", "hu" })
  if win then return win.action, { tile = win.tile, tiles = win.tiles } end
  local gang = prefer({ "ming_gang", "an_gang", "bu_gang", "ti", "pao" })
  if gang then return gang.action, { tile = gang.tile, tiles = gang.tiles } end
  local peng = prefer({ "peng" })
  if peng and math.random() > 0.45 then return "peng", { tile = peng.tile } end
  local chi = prefer({ "chi" })
  if chi and math.random() > 0.55 then return "chi", { tiles = chi.tiles } end
  for _, op in ipairs(ops) do
    if op.action == "pass" then return "pass", {} end
  end
  for _, op in ipairs(ops) do
    if op.action == "discard" then
      local hand = self.game.players[seat].hand
      local tile = hand[math.random(#hand)]
      return "discard", { tile = tile }
    end
  end
  return nil
end

function Room:kick_bots()
  for _ = 1, 24 do
    if self.game.phase == "finished" then return end
    local acted = false
    for i = 0, self.n - 1 do
      local s = self.seats[i]
      if s and s.is_bot then
        local ops = self.game:get_ops(i)
        if #ops > 0 then
          local action, payload = self:bot_act(i, ops)
          if action then
            self.game:apply(i, action, payload)
            acted = true
            break
          end
        end
      end
    end
    if not acted then break end
  end
end

function Room:try_start()
  if self.started then return end
  if not self:all_ready() then return end
  self.started = true
  self.game:start()
  self:kick_bots()
end

function Room:next_round()
  if self.game.phase ~= "finished" then return end
  if not self:all_ready() then return end
  self.started = true
  self.game:start()
  self:kick_bots()
end

function Room:on_ready(seat)
  self:set_ready(seat)
  if self.game.phase == "finished" then
    for i = 0, self.n - 1 do
      if self.seats[i] and self.seats[i].is_bot then
        self.seats[i].ready = true
      end
    end
    self:next_round()
  else
    self:try_start()
  end
end

function Room:on_action(seat, action, payload)
  local err = self.game:apply(seat, action, payload)
  if err then return err end
  if self.game.phase == "finished" then
    for i = 0, self.n - 1 do
      if self.seats[i] then
        if self.seats[i].is_bot then
          self.seats[i].ready = true
        else
          self.seats[i].ready = false
        end
      end
    end
    self.started = false
  else
    self:kick_bots()
  end
  return nil
end

return Room
